src/getting-started.md

```diff
@@ -0,0 +1,470 @@
+for driving CI and building the final release artifacts.
+
+Unfortunately, a proper 2-stage build takes a long time depending on your
+hardware, but it is the only correct way to build everything (e.g. it's what
+the CI and release processes use). **However, in most cases, you can get by
+without a full 2-stage build**. In the following section, we give instructions
+for how to do "the correct thing", but then we also give various tips to speed
+things up.
+
+### Building and Testing `rustc`
+
+To do a full 2-stage build of the whole compiler, you should run this (after
+updating `config.toml` as mentioned above):
+
+```sh
+./x.py build
+```
+
+In the process, this will also necessarily build the standard libraries, and it
+will build `rustdoc` (which doesn't take too long).
+
+To build and test everything:
+
+```sh
+./x.py test
+```
+
+For most contributions, you only need to build stage 1, which saves a lot of time:
+
+```shell
+# Build the compiler (stage 1)
+./x.py build --stage 1
+
+# Subsequent builds
+./x.py build --stage 1 --keep-stage 1
+```
+
+This will take a while, especially the first time. Be wary of accidentally
+touching or formatting the compiler, as `./x.py` will try to recompile it.
+
+**NOTE**: The `--keep-stage 1` will _assume_ that the stage 0 standard library
+does not need to be rebuilt, which is usually true, which will save some time.
+However, if you are changing certain parts of the compiler, this may lead to
+weird errors. Feel free to ask on [zulip][z] if you are running into issues.
+
+This runs a ton of tests and takes a long time to complete. If you are
+working on `rustc`, you can usually get by with only the [UI tests][uitests]. These
+test are mostly for the frontend of the compiler, so if you are working on LLVM
+or codegen, this shortcut will _not_ test your changes. You can read more about the
+different test suites [in this chapter][testing].
+
+[uitests]: ./tests/adding.html#ui
+[testing]: https://rustc-dev-guide.rust-lang.org/tests/intro.html
+
+```
+# First build
+./x.py test --stage 1 src/test/ui
+
+# Subsequent builds
+./x.py test --stage 1 src/test/ui --keep-stage 1
+```
+
+While working on the compiler, it can be helpful to see if the code just
+compiles (similar to `cargo check`) without actually building it. You can do
+this with:
+
+```shell
+./x.py check
+```
+
+This command is really fast (relative to the other commands). It usually
+completes in a couple of minutes on my laptop.
+
+Finally, the CI ensures that the codebase is using consistent style. To format
+the code:
+
+```shell
+# Actually format
+./x.py fmt
+
+# Just check formatting, exit with error
+./x.py fmt --check
+```
+
+*Note*: we don't use stable `rustfmt`; we use a pinned version with a special
+config, so this may result in different style from normal `rustfmt` if you have
+format-on-save turned on. It's a good habit to run `./x.py fmt` before every
+commit, as this reduces conflicts later.
```

Comment:
I think we should add that you can actually use that "pinned version" of `rustfmt` which is built under `build/<target>/stage0/bin/rustfmt`. Knowing this allows to use it directly to format a single file, which `./x.py fmt` cannot do IIRC.

Reply:
I didn't know that, that's super helpful! x.py fmt takes half a minute to run for me, running on only a single file will make it much quicker.

Reply:
Yes, that also saved my life, now I have this as an alias, and I use it in `rust-analyzer` :)